### PR TITLE
Add validation for input objects

### PR DESCRIPTION
**Pull request by mriceron**
Closes https://github.com/ghostdogpr/caliban/issues/231
Not sure how to test the duplicate names rule without building an invalid Input Object by hand.

### CODE REVIEW

File: core/src/main/scala/caliban/validation/Validator.scala [hunk 1]
```diff
@@ -591,6 +592,67 @@ object Validator {
...
 
+  private def validateInputObject(t: __Type): IO[ValidationError, Unit] = {
+    // https://spec.graphql.org/June2018/#IsInputType()
+    def isInputType(t: __Type): Either[__Type, Unit] = t.kind match {
+      case __TypeKind.LIST | __TypeKind.NON_NULL                         => t.ofType.fold[Either[__Type, Unit]](Left(t))(isInputType)
+      case __TypeKind.SCALAR | __TypeKind.ENUM | __TypeKind.INPUT_OBJECT => Right(())
+      case _                                                             => Left(t)
+    }
+
+    def failValidation(msg: String, explanatoryText: String): IO[ValidationError, Unit] =
```
Comment (ghostdogpr) [inline core/src/main/scala/caliban/validation/Validator.scala]: If that's not too much a hassle, could you extract it and use it elsewhere in this file?

File: core/src/main/scala/caliban/validation/Validator.scala [hunk 2]
```diff
@@ -591,6 +592,67 @@ object Validator {
...
+      IO.fail(ValidationError(msg, explanatoryText))
+
+    def validateFields(fields: List[__InputValue]): IO[ValidationError, Unit] =
+      duplicateFieldName(fields).flatMap(_ =>
```
Comment (ghostdogpr) [inline core/src/main/scala/caliban/validation/Validator.scala]: `.flatMap(_ =>` could be just `<*`, which would let you remove `.unit` too

File: core/src/main/scala/caliban/validation/Validator.scala [hunk 3]
```diff
@@ -591,6 +592,67 @@ object Validator {
 
   }
 
+  private def validateInputObject(t: __Type): IO[ValidationError, Unit] = {
+    // https://spec.graphql.org/June2018/#IsInputType()
+    def isInputType(t: __Type): Either[__Type, Unit] = t.kind match {
+      case __TypeKind.LIST | __TypeKind.NON_NULL                         => t.ofType.fold[Either[__Type, Unit]](Left(t))(isInputType)
```
Comment (ghostdogpr) [inline core/src/main/scala/caliban/validation/Validator.scala]: `Option[__Type]` would be simpler no?

**Comment by ghostdogpr** [comment]
Ah also I think it's okay not to test the duplicate names rule since Caliban will only generate something that doesn't break the rule.

File: core/src/main/scala/caliban/validation/Validator.scala (hunk 3 above)
Comment (mriceron) [reply core/src/main/scala/caliban/validation/Validator.scala]: I didn't like it since it's not super clear, by just looking at it, when it's an error or not (having `Some` would be the error state)

**Comment by mriceron** [comment]
> Ah also I think it's okay not to test the duplicate names rule since Caliban will only generate something that doesn't break the rule.

Yeah I thought about that after writing it , I'll remove it

File: core/src/main/scala/caliban/validation/Validator.scala [hunk 4]
```diff
@@ -572,25 +529,78 @@ object Validator {
...
+      case __TypeKind.SCALAR | __TypeKind.ENUM | __TypeKind.INPUT_OBJECT => Right(())
+      case _                                                             => Left(t)
+    }
+
+    def validateFields(fields: List[__InputValue]): IO[ValidationError, Unit] =
+      duplicateFieldName(fields).<*(
```
Comment (ghostdogpr) [inline core/src/main/scala/caliban/validation/Validator.scala]: why not just `<*` without `.` and `()`?

Comment (mriceron) [reply core/src/main/scala/caliban/validation/Validator.scala]: No reason :)
